packages/serverpod_test/lib/src/serverpod_test.dart

```diff
@@ -0,0 +1,415 @@
+          await transactionManager.pushSavePoint();
+        }
+
+        mainTestSession.setAndConfigureServerpodSession(serverpodSession);
+      });
+
+      tearDown(() async {
+        if (rollbackDatabase == RollbackDatabase.afterEach) {
+          await transactionManager.popSavePoint();
+          await transactionManager.pushSavePoint();
+        }
+
+        if (resetTestSessions == ResetTestSessions.afterEach) {
+          for (var testSession in allTestSessions) {
+            await testSession.resetState();
+          }
+        }
+      });
+
+      tearDownAll(() async {
+        if (rollbackDatabase == RollbackDatabase.afterAll ||
+            rollbackDatabase == RollbackDatabase.afterEach) {
+          await transactionManager.cancelTransaction();
+        }
+
+        for (var testSession in allTestSessions) {
+          await testSession.destroy();
+        }
+        allTestSessions.clear();
+
+        await testServerpod.shutdown();
+      });
+
+      testClosure(testServerpod.testEndpoints, mainTestSession);
+    });
+  };
+}
+
+class _TransactionManager {
+  final List<String> _savePointIds = [];
+
+  Transaction? _transaction;
+
+  late Completer _endTransactionScopeCompleter;
+
+  late Session serverpodSession;
+
+  _TransactionManager(this.serverpodSession);
+
+  Future<Transaction> createTransaction() async {
+    if (_transaction != null) {
+      throw StateError('Transaction already exists.');
+    }
+
+    _endTransactionScopeCompleter = Completer();
+    var transactionStartedCompleter = Completer();
+    late Transaction localTransaction;
+
+    unawaited(
+      serverpodSession.db.transaction((newTransaction) async {
+        localTransaction = newTransaction;
+
+        transactionStartedCompleter.complete();
+
+        await _endTransactionScopeCompleter.future
+            .timeout(const Duration(seconds: 60), onTimeout: () async {
+          await newTransaction.cancel();
+        });
+      }),
+    );
+
+    await transactionStartedCompleter.future;
+
+    _transaction = localTransaction;
+
+    return localTransaction;
+  }
+
+  Future<void> cancelTransaction() async {
+    var localTransaction = _transaction;
+    if (localTransaction == null) {
+      throw StateError('No ongoing transaction.');
+    }
+
+    await localTransaction.cancel();
+    _endTransactionScopeCompleter.complete();
+    _transaction = null;
+  }
+
+  Future<void> pushSavePoint() async {
+    if (_transaction == null) {
+      throw StateError('No ongoing transaction.');
+    }
+
+    var savePointId = _getNextSavePointId();
+    _savePointIds.add(savePointId);
+
+    await serverpodSession.db.unsafeExecute(
+      'SAVEPOINT $savePointId;',
+      transaction: _transaction,
+    );
+  }
+
+  String _getNextSavePointId() {
+    var postgresCompatibleRandomString =
+        const Uuid().v4obj().toString().replaceAll(RegExp(r'-'), '_');
+    var savePointId = 'savepoint_$postgresCompatibleRandomString';
```

Comment:
Suggestion: Add an index to the savepoint id so that they are easier to track.

Reply:
Very easy to do, but I thought we said these aren't inspectable? Or do you mean if we need to debug this code in the future?